### PR TITLE
PROPOSAL: add 'conflict-class:'

### PR DESCRIPTION
**Pull request by AltGr**
This might be the missing piece we needed most from `provides:`, but without the problems that come with the other ones

### CODE REVIEW

**Comment by Drup** [comment]
Does this means we can get rid of the difference between `ocaml-base-compiler` and `ocaml-variants` ?

**Comment by AltGr** [comment]
What difference do you mean ?
This would change the way conflicts are defined, i.e. all of `ocaml-system`, `ocaml-core-compiler` and `ocaml-variants` will need to define `conflict-class: "ocaml-core-compiler"`, instead of having the first two define conflicts with the two others, and the last not needing to define any conflict at all.

The big difference is that this solution scales if we increase the number of implementing packages, so this could open the way to having more specific packages instead of `ocaml-variants`, while at the moment all variants are in the same package, with the variant name encoded as the suffix of the version (which doesn't make choosing or conflicting with a given variant easy).

Of course, the virtual `ocaml` package would still need to list the implementors in its `depends:` field (but only once (per version)). Which, although a little cumbersome, may be best for security concerns anyway.

File: src/client/opamMain.ml [hunk 1]
```diff
@@ -2688,6 +2688,287 @@ let clean =
         $switch $all_switches),
   term_info "clean" ~doc ~man
 
+(* BUILD *)
```
Comment (avsm) [inline src/client/opamMain.ml]: This PR includes `opam-build` -- is that intended?

Comment (AltGr) [inline src/client/opamMain.ml]: Indeed, it'll be rebased anyway, now that `opam build` is merged

**Comment by Drup** [comment]
@AltGr So concretely, it means we can add a `metaocaml` separate package, and avoid burying it inside `ocaml-variants.4.04+metaocaml` by just adding the conflict-class (and the dependency in `ocaml`) ? That sounds great.

**Comment by AltGr** [comment]
Exactly. OTOH, when adding a new compiler variant, don't forget to put it in the right `conflict-class:`.
A specific `ocaml-dev` compiler containing all the `+prXXX` versions could be nice as well, as these are not really "variants". It would have started being too cumbersome to get above three or four implementing packages without this.

**Pull request merged**